cdap-master/src/main/java/io/cdap/cdap/master/upgrade/UpgradeJobMain.java

```diff
@@ -0,0 +1,117 @@
+import io.cdap.cdap.client.NamespaceClient;
+import io.cdap.cdap.client.ProgramClient;
+import io.cdap.cdap.client.ScheduleClient;
+import io.cdap.cdap.client.config.ClientConfig;
+import io.cdap.cdap.client.config.ConnectionConfig;
+import io.cdap.cdap.common.NotFoundException;
+import io.cdap.cdap.common.service.Retries;
+import io.cdap.cdap.common.service.RetryStrategies;
+import io.cdap.cdap.common.service.RetryStrategy;
+import io.cdap.cdap.proto.ApplicationRecord;
+import io.cdap.cdap.proto.NamespaceMeta;
+import io.cdap.cdap.proto.ProgramType;
+import io.cdap.cdap.proto.id.ApplicationId;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.id.ScheduleId;
+import io.cdap.cdap.proto.id.WorkflowId;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * Suspends all schedules and stops all programs.
+ * The first parameter to this job should be the hostname, and the second should be the port the
+ * router service is running on. Eg, if the router is running at URI http://your-hostname:11015, then pass in
+ * your-hostname as the first parameter and 11015 as the second.
+ */
+public class UpgradeJobMain {
+
+  private static final int DEFAULT_READ_TIMEOUT_MILLIS = 90 * 1000;
+  private static final String SCHEDULED = "SCHEDULED";
+
+  public static void main(String[] args) {
+    if (args.length != 2) {
+      throw new RuntimeException(
+        String.format("Invalid number of arguments to UpgradeJobMain. Needed 2, found %d", args.length));
+    }
+    ConnectionConfig connectionConfig = ConnectionConfig.builder()
+      .setHostname(args[0])
+      .setPort(Integer.parseInt(args[1]))
+      .setSSLEnabled(false)
+      .build();
+    ClientConfig clientConfig =
+      ClientConfig.builder()
+        .setDefaultReadTimeout(DEFAULT_READ_TIMEOUT_MILLIS)
+        .setConnectionConfig(connectionConfig)
+        .build();
+    RetryStrategy retryStrategy =
+      RetryStrategies.timeLimit(30, TimeUnit.SECONDS,
+                                RetryStrategies.exponentialDelay(10, 500, TimeUnit.MILLISECONDS));
+
+    try {
+      Retries.callWithRetries(
+        () -> {
+          suspendSchedulesAndStopPipelines(clientConfig);
+          return null;
+        }, retryStrategy, e -> e instanceof IOException || e instanceof NotFoundException);
+    } catch (Exception e) {
+      throw new RuntimeException("Failed to prepare instance for upgrade.", e);
+    }
+  }
+
+  private static void suspendSchedulesAndStopPipelines(ClientConfig clientConfig) throws Exception {
+    ApplicationClient applicationClient = new ApplicationClient(clientConfig);
+    ScheduleClient scheduleClient = new ScheduleClient(clientConfig);
+    ProgramClient programClient = new ProgramClient(clientConfig);
+    NamespaceClient namespaceClient = new NamespaceClient(clientConfig);
+
+    for (NamespaceMeta namespaceMeta : namespaceClient.list()) {
```

Comment:
Does the system namespace come back in this list? I seem to remember that it is treated specially and does not.

By default, we run the wrangler and pipeline services in the system namespace and users can also run their own custom apps there. We should be sure to stop them too.

Reply:
Good point, done.